### ISSUE TITLE
GitLab dist URL is wrong for GitLab API v3

### ISSUE THREAD

**Issue by sjadema** [composer/composer#6198]
The dist URL for the GitLab driver generated by `Composer\Repository\Vcs\GitLabDriver::getDist()` uses `$url = $this->getApiUrl().'/repository/archive.zip?sha='.$identifier;` The correct query string parameter is `ref` as opposed to `sha`. Fixing this would enable GitLab repositories to be downloaded as zip packages  instead of being cloned.

This parameters is not documented correctly on the GitLab [docs](https://docs.gitlab.com/ce/api/repositories.html#get-file-archive) as well. A look into the source code for the responsible [controller](https://gitlab.com/gitlab-org/gitlab-ce/blob/v8.17.0/app/controllers/projects/repositories_controller.rb) lets us know is has to be `ref`. An [issue](https://gitlab.com/gitlab-com/gitlab-docs/issues/80) regarding this 'bug' has been submitted.

**Comment by alcohol** [comment]
Well spotted. Can you confirm first on the end of Gitlab if the documentation is wrong, or if the source code is wrong? If you change this now to `ref`, and they actually change their source code to use `sha` in the future, we have to undo your change again.

**Comment by sjadema** [comment]
I left a comment on the other [issue](https://gitlab.com/gitlab-org/gitlab-ce/issues/28689) but they're not that fast it seems. I will keep this issue updated with the status of the other one.

**Comment by sjadema** [comment]
Still no [response](https://gitlab.com/gitlab-org/gitlab-ce/issues/28689). It's is labeled as `Documentation`, `docs-todo` however, so they'll most likely update the documentation opposed to the source code.

**Comment by Seldaek** [comment]
Yeah I think we can merge for now, but please get back to us if you get some conflicting news from them :)

**Comment by sjadema** [comment]
Thanks for fixing this. If anything changes on GitLab's end, i will report back!

**Comment by curry684** [comment]
Given that a [new major API version is around the corner](https://docs.gitlab.com/ee/api/v3_to_v4.html) I think it's pretty safe to rely on them not breaking BC in the EOL API.

**Comment by liweiyi88** [comment]
For Gitlab API V3, it still uses `sha` instead of `ref` in the query string.
We are using composer version `1.4.1`, the dist URL for gitlab is wrong after #6199 was merged.
They are going to move `sha` to `ref` for V4. But it needs some changes for composer to use its V4 api.

**Comment by Seldaek** [comment]
@sjadema did you actually check that ref was working fine with v3 api URLs?

**Comment by sjadema** [comment]
@Seldaek Yes, i did extensive checking with the `sha` parameter and it was not working for `v3`. We are using Composer `1.4.1`  in production right now and it is working fine (in combination with GitLab CE `v9.0.4` and the `ref` parameter). I'll post an example link to an archive in a moment.

**Comment by Seldaek** [comment]
@sjadema Well in that case either @liweiyi88 or you is messing something up.. But those two statements are not compatible :)

**Comment by sjadema** [comment]
@Seldaek, I did some more testing and it seems @liweiyi88 is right. When downloading with the `ref` parameter, GitLab skips the parameter and downloads the `master` branch. I don't know if this behaviour changed between GitLab `8.x` & `9.x` as i did my testing for this issue with GitLab `8.x`.

My initial testing could be flawed as well and maybe i overlooked the differences in repositories. It seems the parameter should be reverted to `sha`. I'll create another issue to revert this change. I do apologize for the inconvenience this might have caused.

**Comment by Seldaek** [comment]
No worries, it happens. I'll make sure to revert before tagging the next bugfix release.

**Comment by axilleas** [comment]
Sorry for the confusion, seems this change missed our docs :/ This was the MR that changed the endpoint https://gitlab.com/gitlab-org/gitlab-ce/merge_requests/9637.

Thanks @sjadema for opening https://gitlab.com/gitlab-org/gitlab-ce/issues/28689.

**Comment by liweiyi88** [comment]
Thanks @sjadema  @Seldaek !
